core/src/avm2/globals/flash/net/sharedobject.rs

```diff
@@ -39,28 +47,444 @@ fn class_init<'gc>(
+        Value::String(s) => Some(AmfValue::String(s.to_string())),
+        Value::Object(o) => {
+            // TODO: Find a more general rule for which object types should be skipped,
+            // and which turn into undefined.
+            if o.as_executable().is_some() {
+                None
+            } else if o.as_display_object().is_some() {
+                Some(AmfValue::Undefined)
+            } else if let Some(array) = o.as_array_storage() {
+                let mut values = Vec::new();
+                recursive_serialize(activation, o, &mut values).unwrap();
+
+                if values.len() == array.length() {
+                    Some(AmfValue::StrictArray(
+                        values.into_iter().map(|elem| elem.value).collect(),
+                    ))
+                } else {
+                    let len = values.len() as u32;
+                    Some(AmfValue::ECMAArray(vec![], values, len))
+                }
+            } else if let Some(date) = o.as_date_object() {
+                date.date_time()
+                    .map(|date_time| AmfValue::Date(date_time.timestamp_millis() as f64, None))
+            } else {
+                let is_object = o
+                    .instance_of()
+                    .map_or(false, |c| c == activation.avm2().classes().object);
+                if is_object {
+                    let mut object_body = Vec::new();
+                    recursive_serialize(activation, o, &mut object_body).unwrap();
+                    Some(AmfValue::Object(
+                        object_body,
+                        Some(ClassDefinition {
+                            name: "".to_string(),
+                            attributes: EnumSet::only(Attribute::Dynamic),
+                            static_properties: Vec::new(),
+                        }),
+                    ))
+                } else {
+                    log::warn!(
+                        "Serialization is not implemented for class other than Object: {:?}",
+                        o
+                    );
+                    None
+                }
+            }
+        }
+    }
+}
+
+/// Serialize an Object and any children to a JSON object
+fn recursive_serialize<'gc>(
+    activation: &mut Activation<'_, 'gc, '_>,
+    obj: Object<'gc>,
+    elements: &mut Vec<Element>,
+) -> Result<(), Error> {
+    let mut last_index = obj.get_next_enumerant(0, activation)?;
+    while let Some(index) = last_index {
+        let name = obj
+            .get_enumerant_name(index, activation)?
+            .coerce_to_string(activation)?;
+        let value = obj.get_property(&QName::dynamic_name(name).into(), activation)?;
+
+        if let Some(value) = serialize_value(activation, value) {
+            elements.push(Element::new(name.to_utf8_lossy(), value));
+        }
+        last_index = obj.get_next_enumerant(index, activation)?;
+    }
+    Ok(())
+}
+
+/// Deserialize a AmfValue to a Value
+fn deserialize_value<'gc>(
```

Comment:
There actually was already a deserialization function [here](https://github.com/ruffle-rs/ruffle/blob/master/core/src/avm2/globals/flash/utils/bytearray.rs#L20-L87). It might be a good idea to merge your implementation with that one, since it implements some things the other does not, and lets maybe move these serialization/deserialization into a different file to avoid confusion.

Reply:
Good catch - I've merged the two implementations, and moved all of the AMF code to a new `amf` module.